psp/src/vram_alloc.rs

```diff
@@ -54,38 +62,55 @@ impl VramMemChunk {
 // TODO: pin?
 #[derive(Debug)]
 pub struct SimpleVramAllocator {
-    offset: u32,
+    offset: AtomicU32,
 }
 
 impl SimpleVramAllocator {
     const fn new() -> Self {
-        Self { offset: 0 }
+        Self {
+            offset: AtomicU32::new(0),
+        }
     }
 
-    // TODO: return a Result instead of panicking
-    pub fn alloc(&mut self, size: u32) -> VramMemChunk {
-        let old_offset = self.offset;
-        self.offset += size;
+    /// Frees all previously allocated VRAM chunks.
+    ///
+    /// This resets the allocator's counter, but does not change the contents of
+    /// VRAM. Since this method requires `&mut Self`, it cannot overlap with any
+    /// previously allocated `VramMemChunk`s since they have the lifetime of the
+    /// `&Self` that allocated them.
+    pub fn free_all(&mut self) {
+        self.offset.store(0, Ordering::Relaxed);
+    }
 
-        if self.offset > self.total_mem() {
+    // TODO: return a Result instead of panicking
+    /// Allocates `size` bytes of VRAM
+    ///
+    /// The returned VRAM chunk has the same lifetime as the
+    /// `SimpleVramAllocator` borrow (i.e. `&self`) that allocated it.
+    pub fn alloc<'a>(&'a self, size: u32) -> VramMemChunk<'a> {
+        let old_offset = self.offset.load(Ordering::Relaxed);
+        let new_offset = old_offset + size;
+        self.offset.store(new_offset, Ordering::Relaxed);
```

Comment:
I see a good old race condition right there.

Reply:
Is this intentional btw? Perhaps this was written as a singlethreaded code?

Reply:
`SimpleVramAllocator` is used as a singleton, so I was treating this like singlethreaded code. The AtomicU32 is only for interior mutability and I went with that over `Cell` to make it easier in case anyone wants to try using the allocator from multiple threads (though that would require a few more changes).

Reply:
But the problem is `SimpleVramAllocator` implements `Sync`, which means you can share references of it between different threads, and function in question (`alloc`) requires just a shared reference, so `alloc` can be simultaneously run on multiple different threads. in that case `alloc` can return two aliasing mem chunks.

Reply:
In bumpalo thread support is provided by a ["herd"](https://crates.io/crates/bumpalo-herd) which is [roughly a `Mutex<Vec<Bump>>`](https://docs.rs/bumpalo-herd/0.1.1/src/bumpalo_herd/lib.rs.html#94) that creates a new bump allocator with its own pool of memory for each thread. Splitting VRAM per thread seems impractical and I don't think there's a safe way to use one bump allocator from multiple threads. So I'd say let's just add `impl !Sync for SimpleVramAllocator` for now.